### ISSUE TITLE
[2017 NIPS Oral] Train longer, generalize better: closing the generalization gap in large batch training of neural networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#74]
<img width="746" alt="image" src="https://user-images.githubusercontent.com/49643709/172014066-91cdc514-d728-430f-b330-694bdafdf4b6.png">

`On Large-Batch Training for Deep Learning: Generalization Gap and Sharp Minima` 논문에서 small batch와 large batch에서 학습한 모델 간의 generalization gap이 있음을 관측함
(해당 논문에서는 sharp minima, large minima 때문이라고 함)

해당 논문은 generalization gap이 왜 일어지는지를 설명하며, 어떻게 극복할 수 있는지 보임

### 1.1. Motivation

Loss의 exact shape를 모르는 complex system에서, statistical physics 모델들은 potential을 random process로 가정함
(in complex systems where we do not know the exact shape of the loss, statistical physics models commonly assume a simpler description of the potential as a random process)

### 1.2. Model : Random walk on a random potential

Physical intuition에서 따와서, model의 weight vector를 loss landscape에서 random walk를 수행하는 particle이라고 볼 수 있음

Momentum term을 붙이는 것은 particle에 inertia가 붙었다고 생각할 수 있음

기존에 high dimension particle이 random potential에서 random walk하는지 연구가 있었으며, main result는 다음과 같음

<img width="777" alt="image" src="https://user-images.githubusercontent.com/49643709/172013381-083f424e-0d54-43c7-bbc6-61c5d67557fc.png">

(asymtotic이란 말은 $x$가 무한히 가면 $x^2 + 5x$를 $x^2$라고 봐도 된다는 느낌)

즉 이야기하고 싶은 바는 model weight가 loss surface에서 움직이는데, 얼만큼 움직였는지를 iteration t의 식으로 표현할 수 있다는 것임

### 1.3. Comparison with empirical results and implications
<img width="765" alt="image" src="https://user-images.githubusercontent.com/49643709/172014053-48f67dd3-516f-4b91-94a2-a782f5e1dc72.png">

<img width="752" alt="image" src="https://user-images.githubusercontent.com/49643709/172014139-f17c7593-dcaa-4ab3-a578-451c048cdc92.png">

저자들은 실험을 통해 alpha가 2임을 찾음 (Figure 2a)

전반적으로 모든 batch size에서 비슷한 logarithmic graph가 관측됨을 확인할 수 있으나, 2가지 다른점이 있음

1. 각 batch size graph는 different slope를 가짐
2. 모든 batch size에서 epoch를 통일하였기에, large batch일수록 iteration이 짧음

저자들은 해당 관측을 토대로 극복 방안을 제시함

### 2.1. Learning rate
<img width="775" alt="image" src="https://user-images.githubusercontent.com/49643709/172016954-76ae01a9-4028-4cb5-bb34-bdb2fe9cca0a.png">

각 batch size마다 slope가 달랐기에, 이를 조정해주기 위해 batch size마다 learning rate를 달리해줌

저자들은 covariance matrix를 이용하여 learning rate를 mini-batch size의 square root만큼 키워야함을 보임

Divergence를 막기 위해 first few iteration에서 gradient를 clip하거나 normalize해줌

해당 식은 per-sample gradient가 sampled mini-batch에 depend하지 않다는 가정하에 나온 것인데, batch normalization은 mini-batch에 dependent함

### 2.2. Ghost Batch Normalization
<img width="759" alt="image" src="https://user-images.githubusercontent.com/49643709/172017276-eb511316-2227-4243-bb9a-6e2e494eebb8.png">

기존의 BN은 주어진 batch size의 batch statistics를 사용함

저자들은 small virtual (ghost) batches의 statistics를 이용하여 이러한 dependency를 없애, generalization error를 줄일 수 있었음

### 2.3. Result
<img width="752" alt="image" src="https://user-images.githubusercontent.com/49643709/172014139-f17c7593-dcaa-4ab3-a578-451c048cdc92.png">

Learning rate을 batch size에 따라 조절하고 GBN을 사용하여 모든 batch size graph가 비슷하게 만들 수 있었음

Constant shift를 관측할 수 있는데, 이는 gradient clipping으로 발생한 것이라고 저자들은 생각함

## 3. Adapting number of weight updates eliminates generalization gap

기존의 학습 방법들은 overfitting을 피하기 위해 training error는 감소하는데 validation error가 그대로면 learning rate를 줄였음
(practitioners decrease the learning rate after validation error appears to reach a plateau, for fear of overfitting)

하지만 저자들은 training error는 감소하는데 validation error가 그대로인 상황에서 같은 learning rate로 계속 optimize를 하면 성능이 증가함을 관측함

이를 통해 저자들은 generalization gap 현상이 batch size보다는 small number of update 때문이라고 생각함

저자들은 batch size에 상관없이 똑같은 iteration을 가지도록 time-frame을 stretch함

## 4. Experiments
<img width="815" alt="image" src="https://user-images.githubusercontent.com/49643709/172017827-507a055d-657f-4046-8b2f-49394a1a977e.png">

- +LR : modifying the learning-rate used for large-batch
- +GBN : using the Ghost-BN method
- +RA : training regime is adapted
